internal/webhooks/vmware/vspheremachinetemplate.go

```diff
@@ -0,0 +1,108 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vmware is the package for webhooks of vmware resources.
+package vmware
+
+import (
+	"context"
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	vmwarev1 "sigs.k8s.io/cluster-api-provider-vsphere/apis/vmware/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/services/vmoperator"
+)
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-vmware-infrastructure-cluster-x-k8s-io-v1beta1-vspheremachinetemplate,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=vmware.infrastructure.cluster.x-k8s.io,resources=vspheremachinetemplates,versions=v1beta1,name=validation.vspheremachinetemplate.vmware.infrastructure.cluster.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+// VSphereMachineTemplateWebhook implements a validation webhook for VSphereMachineTemplate.
+type VSphereMachineTemplateWebhook struct{}
+
+var _ webhook.CustomValidator = &VSphereMachineTemplateWebhook{}
+
+func (webhook *VSphereMachineTemplateWebhook) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&vmwarev1.VSphereMachineTemplate{}).
+		WithValidator(webhook).
+		Complete()
+}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *VSphereMachineTemplateWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	vSphereMachineTemplate, ok := obj.(*vmwarev1.VSphereMachineTemplate)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a VSphereMachineTemplate but got a %T", obj))
+	}
+	return webhook.validate(ctx, nil, vSphereMachineTemplate)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *VSphereMachineTemplateWebhook) ValidateUpdate(ctx context.Context, _, newRaw runtime.Object) (admission.Warnings, error) {
+	vSphereMachineTemplate, ok := newRaw.(*vmwarev1.VSphereMachineTemplate)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a VSphereMachineTemplate but got a %T", newRaw))
+	}
+	return webhook.validate(ctx, nil, vSphereMachineTemplate)
+}
+
+func (webhook *VSphereMachineTemplateWebhook) validate(_ context.Context, _, newVSphereMachineTemplate *vmwarev1.VSphereMachineTemplate) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+
+	// Validate namingStrategy
+	namingStrategy := newVSphereMachineTemplate.Spec.Template.Spec.NamingStrategy
+	if namingStrategy != nil &&
+		namingStrategy.Template != nil {
+		name, err := vmoperator.GenerateVirtualMachineName("machine", namingStrategy)
```

Comment:
Optional: I was wondering if we have a way to bubble up a warning for the truncating case, when passing a long name, but that's not straight forward.

Reply:
Good question, see: https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/3099#discussion_r1671947310

Reply:
Good enough 👍